File: src/segmentation.cpp
#include "segmentation.h"

using namespace notesegmentation;

RTSegmentation::RTSegmentation() : MIN_ONSET_GAP_MS(200) {
    _odf = new PeakAmpDifferenceODF();
    _od = new RTOnsetDetection();

    _freqs = NULL;
    _window = NULL;
    _fft_in = NULL;
    _fft_out = NULL;

    _sampling_rate = 44100;
    hop_size(512);
    frame_size(512);

    _min_onset_gap = (_sampling_rate * MIN_ONSET_GAP_MS) / 1000;
    _current_onset_gap = 0;
    _current_region = NONE;

    _peak_rms = 0.f;
    _peak_amp = 0.f;

    _n_prev_odf_values = 2;
    _n_prev_amp_values = 3;
    _n_prev_amp_ma_values = 5;
    _n_amp_mean_values = 3;
    _n_prev_centroid_values = 2;
    _prev_odf_values = new sample[_n_prev_odf_values];
    _prev_amp_values = new sample[_n_prev_amp_values];
    _prev_amp_ma_values = new sample[_n_prev_amp_ma_values];
    _prev_centroid_values = new sample[_n_prev_centroid_values];
    memset(_prev_odf_values, 0.0, sizeof(sample) * _n_prev_odf_values);
    memset(_prev_amp_values, 0.0, sizeof(sample) * _n_prev_amp_values);
    memset(_prev_amp_ma_values, 0.0, sizeof(sample) * _n_prev_amp_ma_values);
    memset(_prev_centroid_values, 0.0,
           sizeof(sample) * _n_prev_centroid_values);

    _n_frames = 0;
    _prev_centroid_cma = 0.f;
}

RTSegmentation::~RTSegmentation() {
    if(_odf) {
        delete _odf;
        _odf = NULL;
    }
    if(_od) {
        delete _od;
        _od = NULL;
    }
    if(_window) {
        delete [] _window;
        _window = NULL;
    }
    if(_freqs) {
        delete [] _freqs;
        _freqs = NULL;
    }

    if(_fft_in) {
        fftw_free(_fft_in);
        _fft_in = NULL;
    }
    if(_fft_out) {
        fftw_free(_fft_out);
        _fft_out = NULL;
    }
    fftw_destroy_plan(_plan);

    if(_prev_odf_values) {
        delete [] _prev_odf_values;
        _prev_odf_values = NULL;
    }
    if(_prev_amp_values) {
        delete [] _prev_amp_values;
        _prev_amp_values = NULL;
    }
    if(_prev_amp_ma_values) {
        delete [] _prev_amp_ma_values;
        _prev_amp_ma_values = NULL;
    }
    if(_prev_centroid_values) {
        delete [] _prev_centroid_values;
        _prev_centroid_values = NULL;
    }
}

// Clear all stored values so ready for new note
void RTSegmentation::reset() {
    _odf->reset();
}

int RTSegmentation::frame_size() {
    return _frame_size;
}

void RTSegmentation::frame_size(int frame_size) {
    _frame_size = frame_size;
    _num_bins = (_frame_size / 2) + 1;

    if(_freqs) {
        delete [] _freqs;
    }
    _freqs = new sample[_num_bins];
    sample base_freq = (sample)_sampling_rate / (sample)_frame_size;
    for(int bin = 0; bin < _num_bins; bin++) {
        _freqs[bin] = bin * base_freq;
    }

    if(_window) {
        delete [] _window;
    }
    _window = new sample[_frame_size];
    windows::hamming(_frame_size, _window);

    if(_fft_in) {
        fftw_free(_fft_in);
    }
    _fft_in = (sample*) fftw_malloc(sizeof(sample) * _frame_size);

    if(_fft_out) {
        fftw_free(_fft_out);
    }
    _fft_out = (fftw_complex*) fftw_malloc(sizeof(fftw_complex) * _num_bins);

    fftw_destroy_plan(_plan);
	_plan = fftw_plan_dft_r2c_1d(_frame_size, _fft_in,
                                 _fft_out, FFTW_ESTIMATE);

    _odf->set_frame_size(_frame_size);
}

int RTSegmentation::hop_size() {
    return _hop_size;
}

void RTSegmentation::hop_size(int hop_size) {
    _hop_size = hop_size;
    _odf->set_hop_size(hop_size);
}

// Return the spectral centroid for a given frame of audio
sample RTSegmentation::spectral_centroid(int n, sample* audio) {
    if(n != _frame_size) {
        throw Exception(std::string("Size of spectral centroid not ") +
                        std::string("equal to segmentation frame size"));
        return 0.f;
    }

    sample centroid = 0.f;
    sample freqs_amps_sum = 0.f;
    sample amps_sum = 0.f;
    sample amp = 0.f;

    memcpy(_fft_in, &audio[0], sizeof(sample) * _frame_size);
    windows::window(_frame_size, _window, _fft_in);
    fftw_execute(_plan);

    for(int bin = 0; bin < _num_bins; bin++) {
        amp = sqrt((_fft_out[bin][0] * _fft_out[bin][0]) +
                   (_fft_out[bin][1] * _fft_out[bin][1]));
        amps_sum += amp;
        freqs_amps_sum += _freqs[bin] * amp;
    }

    if(amps_sum > 0.f) {
        centroid = freqs_amps_sum / amps_sum;
    }

    return centroid;
}

int RTSegmentation::segment(int n, sample* audio) {
    if(n != _frame_size) {
        throw Exception(std::string("Size of audio frame not ") +
                        std::string("equal to segmentation frame size"));
        return NONE;
    }

    sample odf_value = _odf->process_frame(_frame_size, audio);
    sample centroid = spectral_centroid(_frame_size, audio);
    sample centroid_cma = util::cumulative_moving_average(
        _n_frames, centroid, _prev_centroid_cma
    );

    sample rms_value = rms(_frame_size, audio);
    util::rotate(_n_prev_amp_values, _prev_amp_values, rms_value);

    sample rms_ma_value = mean(_n_amp_mean_values, _prev_amp_values);

    for(int i = 0; i < _frame_size; i++) {
        if(fabs(audio[i]) > _peak_amp) {
            _peak_amp = fabs(audio[i]);
        }
    }
    if(rms_ma_value > _peak_rms) {
        _peak_rms = rms_ma_value;
    }

    bool is_odf_minima = util::is_minima(odf_value, 1, _prev_odf_values);
    bool is_rms_maxima = util::is_maxima(rms_ma_value, 1, _prev_amp_ma_values);

    util::rotate(_n_prev_odf_values, _prev_odf_values, odf_value);
    util::rotate(_n_prev_amp_ma_values, _prev_amp_ma_values, rms_ma_value);

    // update onset and offset, as they can only last for 1 frame
    if(_current_region == ONSET) {
        _current_region = ATTACK;
    }
    else if(_current_region == OFFSET) {
        _current_region = NONE;
    }

    _current_onset_gap += _frame_size;

    if(_od->is_onset(odf_value)) {
        // if not too close to a previous onset, start a new note attack
        if(_current_onset_gap >= _min_onset_gap) {
            _current_onset_gap = 0;
            _current_region = ONSET;
        }
    }

    if(_current_region == ATTACK) {
        // transient/attack: from onset until next minima in odf,
        // unless amp local max reached first
        if(is_odf_minima || is_rms_maxima) {
            _current_region = SUSTAIN;
        }
    }
    else if(_current_region == SUSTAIN) {
        // release: 3 consecutive frames with decreasing energy and
        //          spectral centroid below average
        if((rms_ma_value <= 0.8 * _peak_rms &&
            util::decreasing(_n_prev_amp_ma_values, _prev_amp_ma_values) &&
            centroid < centroid_cma) ||
           (rms_ma_value <= 0.33 * _peak_rms)) {
            _current_region = RELEASE;
        }
    }
    else if(_current_region == RELEASE) {
        // offset: point (after sustain) at which envelope <=-60db
        //         below peak value
        sample offset_threshold = 0.001 * _peak_amp;
        if(rms_value <= offset_threshold) {
            _current_region = OFFSET;
        }
    }

    _n_frames += 1;
    _prev_centroid_cma = centroid_cma;

    return _current_region;
}

File: notesegmentation/segmentation.py
import numpy as np
import modal
import modal.onsetdetection as od
import amplitude_envelopes as ae
import util


class NoOnsetsFound(Exception):
    pass


def odf(audio, metadata, return_odf=False):
    frame_size = int(metadata.get('odf_frame_size', 2048))
    hop_size = int(metadata.get('odf_hop_size', 512))
    sampling_rate = int(metadata.get('sampling_rate', 44100))

    o = modal.PeakAmpDifferenceODF()
    o.set_frame_size(frame_size)
    o.set_hop_size(hop_size)
    o.set_sampling_rate(sampling_rate)

    onsets = []
    odf_values = []
    onset_det = od.RTOnsetDetection()

    p = 0
    while p <= len(audio) - frame_size:
        frame = audio[p:p + frame_size]
        odf_value = o.process_frame(frame)
        odf_values.append(odf_value)
        det_results = onset_det.is_onset(
            odf_value, o.max_odf_value(), return_threshold=True
        )
        if det_results[0]:
            onsets.append(p)
        p += hop_size

    if not return_odf:
        return onsets
    else:
        return (onsets, np.array(odf_values))


def spectral_centroid(audio, metadata):
    frame_size = int(metadata.get('spectral_centroid_frame_size', 512))
    hop_size = int(metadata.get('spectral_centroid_hop_size', 512))
    sampling_rate = int(metadata.get('sampling_rate', 44100))

    num_bins = (frame_size / 2) + 1
    freqs = np.linspace(0.0, float(sampling_rate) / 2, num_bins)
    sc = []

    p = 0
    while p <= len(audio) - frame_size:
        frame = audio[p:p + frame_size]
        centroid = 0.0

        window = np.hamming(frame_size)
        f = np.fft.rfft(frame * window)
        magnitudes = abs(f)

        if sum(magnitudes):
            centroid = sum(freqs * magnitudes) / sum(magnitudes)

        sc.append(centroid)
        p += hop_size

    return np.array(sc)


def _effort_times(s, thresholds):
    t = np.zeros(len(thresholds))
    current_threshold = 0
    for i in range(len(s)):
        if s[i] >= thresholds[current_threshold]:
            t[current_threshold] = i
            current_threshold += 1
            if current_threshold >= len(thresholds):
                break
    # this shouldn't happen, but just in case...
    if current_threshold < len(thresholds):
        t[current_threshold:len(thresholds)] = len(s)
    return t


def _efforts(s, thresholds):
    effort_times = _effort_times(s, thresholds)
    efforts = np.diff(effort_times)
    w = np.mean(efforts)
    M = len(efforts[efforts > w])  # no. efforts greater than mean
    return efforts, w, M


def cbr(audio, metadata, verbose=False):
    """
    Implementation of the automatic note segmentation given in:

        Caetano, M., Burred, J. J., Rodet, X.
        Automatic Segmentation of the Temporal Evolution of Isolated Acoustic
        Musical Instrument Sounds Using Spectro-Temporal Cues.
        In Proc. DAFx 2010

    Returns a list of dictionaries (one for each note) with the following keys:
    * onset
    * end_attack
    * sustain
    * release
    * offset
    """
    if verbose:
        print 'Segmenting %s with the Caetano, Burred, Rodet method' % \
            metadata.get('name', 'unknown audio file')

    notes = []

    onsets = odf(audio, metadata)
    if not onsets:
        raise NoOnsetsFound()
    onsets = [onsets[0]]

    env = ae.tae(np.abs(audio))
    sc = spectral_centroid(audio, metadata)

    for onset_number, onset in enumerate(onsets):
        # if time to prev onset is < 200 ms, ignore this onset
        note_duration = onset - onsets[onset_number - 1]
        max_note_duration = (metadata['sampling_rate'] * 200) / 1000
        if onset_number > 0 and note_duration < max_note_duration:
                if verbose:
                    print 'Warning: Detected note duration is too short',
                    print '(%d to %d, %d samples). Skipping.' % (
                        onsets[onset_number - 1], onset, note_duration
                    )
                continue

        boundaries = {
            'onset': onset,
            'end_attack': onset,
            'sustain': onset,
            'release': onset,
            'offset': len(audio)
        }

        # offset: last point TAE has same energy as onset
        n = 512
        onset_energy = np.sum(env[np.max(onset - n, 0):onset] ** 2)
        audio_pos = onset + (metadata['sampling_rate'] * 200 / 1000)
        while audio_pos <= len(audio) - n:
            frame = env[audio_pos:audio_pos + n]
            energy = np.sum(frame ** 2)
            if energy <= onset_energy:
                break
            audio_pos += n
        boundaries['offset'] = audio_pos

        # beginning of sustain / end of attack
        # using a modified version of Peeters' efforts method
        max_transient = min(onset + (metadata['sampling_rate'] * 500) / 1000,
                            len(env) - onset)
        if onset == max_transient:
            if verbose:
                print 'Warning: onset detected at end of signal, ignoring'
            continue
        max_env = np.max(env[onset:max_transient])
        max_env_loc = np.argmax(env[onset:max_transient]) + onset
        effort_thresholds = np.linspace(0, max_env, 10)
        effort_times = \
            _effort_times(env[onset:max_env_loc], effort_thresholds) + onset
        efforts = np.diff(effort_times)
        w = np.mean(efforts)
        M = len(efforts[efforts > w])  # no. efforts greater than mean

        # start of sustain: first point at which effort > M * w
        sustain = max_env_loc
        for i, effort in enumerate(efforts):
            if effort > M * w:
                sustain = effort_times[i]
        boundaries['sustain'] = sustain

        # end of attack: first local minima (in spectral centroid)
        # between onset and sustain
        sc_minima = util.next_minima(sc, onset / 512) * 512
        boundaries['end_attack'] = min(sustain, sc_minima)

        # start of release:  using a modified version of Peeters'
        # efforts method (in reverse)
        max_env = np.max(env[onset:boundaries['offset']])
        max_env_loc = np.argmax(env[onset:boundaries['offset']]) + onset
        effort_thresholds = np.linspace(0, max_env, 10)
        effort_times = _effort_times(env[boundaries['offset']:onset:-1],
                                     effort_thresholds)
        efforts = np.diff(effort_times)
        w = np.mean(efforts)
        M = len(efforts[efforts > w])

        # first point at which effort > M * w
        release = max_env_loc
        for i, effort in enumerate(efforts):
            if effort > M * w:
                release = boundaries['offset'] - effort_times[i]
        boundaries['release'] = release

        notes.append(boundaries)
    return notes


def rtsegmentation(audio, metadata, verbose=False):
    """
    Implementation of real-time automatic note segmentation, as described in:
        Glover, J., Lazzarini, V., and Timoney, J.
        Real- time segmentation of the temporal evolution of musical sounds
        Proceedings of the Acoustics 2012 Hong Kong Conference

    Returns a list of dictionaries (one for each note) with the following keys:
    * onset
    * sustain
    * release
    * offset
    """
    if verbose:
        print 'Segmenting %s with the Glover, Lazzarini and Timoney method' % \
            metadata.get('name', 'unknown audio file')

    notes = []

    onsets, o = odf(audio, metadata, return_odf=True)
    if not onsets:
        raise NoOnsetsFound()
    onsets = [onsets[0]]

    odf_hop_size = metadata.get('odf_hop_size', 512)
    env_hop_size = metadata.get('env_hop_size', 512)

    env = ae.rms_frame(audio, n=env_hop_size)
    env_no_ma = ae.rms_frame(audio, n=env_hop_size, m=1)
    sc = spectral_centroid(audio, metadata)
    centroid_cma = util.cumulative_moving_average(sc)

    for onset_number, onset in enumerate(onsets):
        # if time to prev onset is < 200 ms, ignore this onset
        note_duration = onset - onsets[onset_number - 1]
        max_note_duration = (metadata['sampling_rate'] * 200) / 1000
        if onset_number > 0 and note_duration < max_note_duration:
                if verbose:
                    print 'Warning: Detected note duration is too short',
                    print '(%d to %d, %d samples). Skipping.' % (
                        onsets[onset_number - 1], onset, note_duration
                    )
                continue

        boundaries = {'onset': onset,
                      'sustain': onset,
                      'release': len(audio),
                      'offset': len(audio)}

        onset_frame = onset / odf_hop_size

        # transient/attack: from onset until next minima in odf,
        #                   unless amp local max reached first
        boundaries['sustain'] = min(
            util.next_maxima_rt(env, onset_frame) * env_hop_size,
            util.next_minima_rt(o, onset_frame) * odf_hop_size
        )

        # release: 3 consecutive frames with decreasing energy and
        #          spectral centroid below average
        r = util.next_maxima(env, onset_frame)
        peak_amp = env[r]
        for i in range(r, len(env) - 1):
            peak_amp = max(peak_amp, env[i])
            is_release = ((env[i] <= (0.8 * peak_amp) and
                           util.decreasing(env, i, 5) and
                           sc[i] < centroid_cma[i]) or
                          env[i] <= 0.33 * peak_amp)
            if is_release:
                boundaries['release'] = i * env_hop_size
                break

        # offset: point (after sustain) at which envelope <=-60db
        #         below peak value
        peak_amp = np.max(np.abs(audio))
        peak_location = np.argmax(np.abs(audio)) / env_hop_size
        rt60 = (10 ** -3) * peak_amp
        offset = peak_location * env_hop_size
        for i in range(peak_location, len(env_no_ma)):
            if env_no_ma[i] <= rt60:
                offset = i * env_hop_size
                break
        boundaries['offset'] = offset

        notes.append(boundaries)
    return notes
